### PR TITLE
Memmap: Replace GetPointer with GetSpanForAddress

### PR DESCRIPTION
**Pull request by JosJuice**
To ensure memory safety, callers of GetPointer have to perform a bounds check. But how is this bounds check supposed to be performed? GetPointerForRange contained one implementation of a bounds check, but it was cumbersome, and it also isn't obvious why it's correct.

To make doing the right thing easier, the first commit of this PR changes GetPointer to return a span that tells the caller how many bytes it's allowed to access. Then, the second commit adds bounds checks to two of the three callers of GetPointer. The third caller, VertexLoader, will have to wait due to its tight optimization.

### CODE REVIEW

File: Source/Core/Core/HW/Memmap.cpp [hunk 1]
```diff
@@ -405,7 +405,7 @@ u8* MemoryManager::GetPointerForRange(u32 address, size_t size) const
 
   if (span.data() == nullptr)
   {
-    // The address isn't in a valid memmory region.
+    // The address isn't in a valid memory region.
```
Comment (Tilka) [inline Source/Core/Core/HW/Memmap.cpp:408]: Do you mind moving this to the first commit?^^;

Comment (JosJuice) [reply Source/Core/Core/HW/Memmap.cpp:408]: Oh, whoops

**Comment by JosJuice** [comment]
It seems that this PR breaks the decoding of C8 textures in the software backend, but I can't find anything wrong when re-reading the code... Maybe someone with more experience of the graphics backends can find something when reviewing the code?

**Comment by Tilka** [comment]
I fixed three instances of an issue but there might be more. Note that FifoCI is currently down.

**Comment by JosJuice** [comment]
Ah, I need to multiply the index by `sizeof(u16)` when replacing indexing a `u16*` with calling `SafeSpanRead<u16>`! Yes, that makes sense. I checked all the calls to `SafeSpanRead`, and the three calls you caught seem to be the only ones that needed changing.

**Pull request merged**